1. Obtain the sequencing reads in the **FASTQ** file format (***$sample\*.fq*** for not compressed or ***$sample\*.fq.gz*** for gzip compressed).

2. Perform quality control on the **FASTQ** data using `fastp`, including trimming adapter contamination and filtering out low-quality reads (phred score < 30 and trimmed length < 60).

   For single-end data (not compressed):

   `fastp -i $sample.fq -o $sample.out.fq -q 30 -u 50 -n 5 -l 60`

   For paired-end data (gzip compressed):

   `fastp -i $sample.R1.fq.gz -I $sample.R2.fq.gz -o $sample.out.R1.fq.gz -O $sample.out.R2.fq.gz -q 30 -u 50 -n 5 -l 60`

3. Index the reference genome (***$ref.fasta***) and map the trimmed reads to the reference genome using `bwa-mem2`. The mapping results are saved in the **Sequence Alignment Map (SAM)** file format (***$sample.sam***). The alignments in the **SAM** file should be sorted and written to the compressed binary version of a SAM file (**BAM**) using `samtools` to save space.

   First, the reference genome should be indexed:

   `bwa-mem2 index $ref.fasta`

   Then, map the trimmed reads to the indexed reference genome. For single-end data:

   `bwa-mem2 mem $ref.fasta $sample.out.fq.gz > $sample.sam`

   For paired-end data:

   `bwa-mem2 mem $ref.fasta $sample.out.R1.fq.gz $sample.out.R2.fq.gz > $sample.sam`

   Sort the alignments in the **SAM** file and write the sorted output to the BAM file. The unmapped reads can be filtered with the `-q` parameter. Code:

   `samtools view -q 1 -b $sample.sam | samtools sort - -o $sample.sorted.bam`

   The above mapping and sorting commands can be combined into one using the Linux pipe:

   `bwa-mem2 mem $ref.fasta $sample.out.R1.fq.gz $sample.out.R2.fq.gz | samtools view -q 1 -b | samtools sort - -o $sample.sorted.bam`

   Remove the duplicate reads in the **BAM** file using `sambamba`:

   `sambamba markdup -t 10 -p -r $sample.sorted.bam $sample.sorted.RM.bam`

4. Based on the mapping results, remove samples with low mapping depth (> 10Ã—) and insufficient genome coverage (> 95%).

5. There are two commonly used methods for calling variations into the **variant call format (VCF)** file and filtering out the low-quality sites, including `samtools/bcftools` and the `Genome Analysis Toolkit (GATK)`.

   For `samtools/bcftools`, assuming that the location of each **BAM** file has been written to the **$BAM.paths**. The codes are as below:

   ```shell
   # Generate the gzip compressed VCF file containing genotype likelihoods for one or more BAM file(s).
   
   bcftools mpileup -Ou -A -a FORMAT/AD,FORMAT/DP -f $ref.fasta -b $BAM.paths | bcftools call -mv -O z -o $all.vcf.gz --threads 8
   
   # Sites with a phred-scaled quality score (QUAL) of less than 30, an average read depth (DP) of less than 25 and a mapping quality of less than 30 are marked as FILTERED. SNPs within 3 base pairs of an indel or other variant type are marked with SnpGap.
   
   bcftools filter -s FILTERED --SnpGap 3 --set-GTs . -O v --include "%QUAL>=30 && AVG(FMT/DP)>=25 && MQ>=30" $all.vcf.gz > $all.Marked.vcf
   
   # Set the genotype of the failed site (DP less than 3) to a missing value (.).
   
   bcftools filter -S . -e 'FMT/DP<3' $all.Marked.vcf > $all.DPmarked.vcf
   
   # Keep the filtered biallelic sites in $all.final.vcf.
   
   bcftools view -m2 -M2 -v snps -O v -o $all.Biallelic.vcf $all.DPmarked.vcf && grep -v "SnpGap" $all.Biallelic.vcf > $all.final.vcf
   ```

   For the GATK pipeline, a **genomic VCF (GVCF)** is created for each sample to record each site in the file for joint analysis of a cohort in subsequent steps. The codes are as below:

   ```shell
   # Generate the gzip compressed GVCF file for each sample.
   
   gatk --java-options "-Xms8G -Xmx16G" HaplotypeCaller -R $ref.fasta -I $sample.sorted.RM.bam -O $sample.g.vcf.gz -ERC GVCF
   
   # Write all the GVCF file locations to the $GVCF.paths file. Combine the GVCF files into one gzip compressed GVCF file.
   
   gatk --java-options "-Xms8G -Xmx16G" CombineGVCFs -R $ref.fasta -V $GVCF.paths -O $all.g.vcf.gz
   
   # Perform joint genotyping on the combined GVCF file and write the genotyped information to a gzip compressed VCF file.
   
   gatk --java-options "-Xms8G -Xmx16G" GenotypeGVCFs -R $ref.fasta -V $all.g.vcf.gz -O $all.genotype.vcf.gz
   
   # Select SNPs from the original VCF file and perform the hard-filtering.
   
   gatk --java-options "-Xms8G -Xmx16G" SelectVariants -select-type SNP -V $all.genotype.vcf.gz -O $all.snp.vcf.gz
   
   gatk --java-options "-Xms8G -Xmx16G" VariantFiltration -R $ref.fasta -V $all.snp.vcf.gz --filter-expression "QUAL < 30.0 || QD < 2.0 || MQ < 40.0 || FS > 60.0 || SOR > 3.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0" --filter-name SNP_filter -O $all.snp.filter.vcf.gz
   
   gatk --java-options "-Xms8G -Xmx16G" SelectVariants -R $ref.fasta -V $all.snp.filter.vcf.gz --exclude-filtered --restrict-alleles-to BIALLELIC -O $all.snp.pass.vcf.gz
   ```

6. Filter sites with minor allele frequency (MAF) less than 0.005 to remove the less informative sites including singletons, and also remove isolates with more than 20% missing sites using `VCFtools`. The passed records are written to the ***$all.maf.miss.recode.vcf*** file. Code:

   `vcftools --gzvcf $all.snp.pass.vcf.gz --max-missing 0.8 --maf 0.005 --out $all.maf.miss --recode --recode-INFO-all`

7. Evaluate the multiplicity of infection (MOI) of each genome using the `moimix` and `SeqArray` R packages by running commands below:

   ```R
   # Load the required packages.
   
   library(moimix)
   
   library(SeqArray)
   
   # Load the VCF file.
   
   vcf_file <- "$all.maf.miss.recode.vcf"
   
   # Convert the VCF file to GDS format for Fws calculation.
   
   seqVCF2GDS(vcf_file, "$Input.gds")
   
   my_vcf <-seqOpen("All.Cp.OG.gds")
   
   # Calculate the Fws values for each sample and write the output to the specified file.
   
   fws_all <- getFws(my_vcf)
   
   write.table(fws_all, file = "$output.Fws.txt", quote = F, row.names = F, sep = "\t")
   ```

   Finally, remove the potential multiple infections (*Fws* < 0.95) using `VCFtools`. Code:

   `cat $output.Fws.txt | awk '{if ($1<0.95) print $2}' > Multi_Infec.ids && vcftools --vcf $all.maf.miss.recode.vcf --remove Multi_Infec.ids --out $final.vcf --recode --recode-INFO-all`
