### 1 **Phylogenetic analysis**

1. Create alignments by concatenating SNPs in the VCF file using the `vcf2phylip.py` script. By default, the output would be in PHYLIP format named ***$SNPaln.min4.phy***. We can use the `-f` argument to write a FASTA format named ***$SNPaln.min4.fasta***. Code:

   `python vcf2phylip.py -i $final.vcf --output-prefix $SNPaln -f`

2. Build phylogenetic tree using `IQ-TREE` with 1000 bootstraps. Substitution models would be automatically selected using ModelFinder Plus (MFP) and optimized with the option `-bnni`. Code:

   `iqtree2 -s $SNPaln.min4.fasta -m MFP+ASC -B 1000 -bnni -T AUTO --prefix $tree`

3. Submit the output file ***$tree.contree*** to the iTOL webserver for tree visualization and annotation.

### 2 **Principle component analysis (PCA)**

Perform the PCA with unlinked SNPs (linkage disequilibrium < 0.4 in 10-kb window) using the `SNPRelate` R package by running commands below:

```R
# Load the required packages.

library(gdsfmt)

library(SNPRelate)

# Load the VCF file and convert it to the GDS file format.

vcf.fn <- "$final.vcf"

snpgdsVCF2GDS(vcf.fn, "$input.gds", method="biallelic.only")

genofile <- snpgdsOpen("$input.gds")

# Keep the unlinked sites for analysis.

snpset <- snpgdsLDpruning(genofile, ld.threshold=0.4, slide.max.bp = 10000)

snpset.id <- unlist(snpset)

pca <- snpgdsPCA(genofile, snp.id=snpset.id, num.thread=2)

# Write the PCA results to files, including the eigenvectors, eigenvalues, and SNPs used. The data frames can be plotted using the ggplot2 package.

tab <- data.frame(sample.id = pca$sample.id,
                  EV1 = pca$eigenvect[,1], EV2 = pca$eigenvect[,2],
                  EV3 = pca$eigenvect[,3], stringsAsFactors = FALSE)

write.table(tab, file = "$pca.features", quote = F, sep = "\t")

write.table(pca$eigenval, file = "$pca.eigenval", quote = F, sep = "\t")

write.table(pca$snp.id, file = "$pca.usedSNPs", quote = F, sep = "\t")
```

### 3 **Model-based clustering**

1. Extract the same subset of SNPs used in PCA.

2. Transform the VCF file to the PLINK PED format using `VCFtools` and create the BED file for Model-based clustering:

   ```shell
   vcftools --vcf $pca.used.vcf --plink --out $structure
   plink --noweb --file $structure --make-bed --out $structure
   ```

   Infer the population structure using `fastStructure` with K values from 2 to 15. Then choose the appropriate number of model components to explain the structure. The commands are as below:

   ```shell
   for i in `seq 2 15`
   do
   python structure.py --input=$structure --full --seed=100 --output=$fastStructure -K ${i}
   done
   
   python chooseK.py --input=$fastStructure
   ```

3. The Q-matrix results (***$fastStructure.k.Q***) can be imported into R for plotting using the `*barplot()* `function. The results can also be submitted to the CLUMPAK web server.

