### 1 **Linkage disequilibrium decay**

1. Extract the variants for each chromosome using `VCFtools`:

   `for i in {1..8};do vcftools --vcf $final.vcf --chr ${i} --out Chr${i} --recode;done`

2. Calculate LD decay for group A by chromosome using `PopLDdecay`:

   `for i in {1..8};do PopLDdecay -InVCF Chr${i}.recode.vcf -Subpop $A.ids -OutStat A${i}.stat.gz -MaxDist 150; done`

3. Draw the LD decay figure for this group A with multiple chromosomes using the perl script `Plot_OnePop.pl` in the *bin* folder of PopLDdecay. Code:

   `ls A${i}.stat.gz > A_PlotInput.list && perl bin/Plot_OnePop.pl -inList A_PlotInput.list -output FigA -maxX 150 -bin1 20 -bin2 100 -break 50 -keepR `

4. For multiple populations, first draw the LD decay for each population as above. Then write the file path of each population result (*\*.bin*) and the population flag into ***Multipop.bin.list***, such as:

   > FigA.bin  A
   >
   > FigB.bin  B

   Draw the figure using the perl script *Plot_MultiPop.pl* in the *bin* folder of PopLDdecay. Code:

   `perl bin/Plot_MultiPop.pl -inList Multipop.bin.list -output Compare`

### 2 Phylogenetic network

Create alignments by concatenating SNPs in the VCF file using the `vcf2phylip.py` script as described in 2.Population structure analysis. Open the `SplitsTree` app and import the alignments. By default, the `SplitsTree` app computes *P-distances* and then runs *Neighbor Net* to obtain a split network. The existence of connected networks that differ from phylogenetic trees would indicate recombination events between subclasses.

### 3 **ABBA-BABA statistics**

1. Prepare a text file with one individual per row and a tab separating the name of the individual from the name of the population to which it belongs. Calculate the *D* and *f4-ratio* statistics for all possible trios of populations using the Dtrios of `Dsuite`. Code:

   `Dsuite Dtrios $final.vcf SETS.txt`

2. For *D* statistics and *fd* admixture estimates across the genome, use the Python script `ABBABABAwindows.py` from genomics_general. The input file should also be in the same file format as described in 3.3. At least three populations (A, B and C) and one outgroup (D) should be included in the ***$popfile***. Code:

   `python ABBABABAwindows.py -g $input.geno -f phased -o $output.csv -w 25000 -m 100 -s 10000 -P1 A -P2 B -P3 C -O D -T 10 --minData 0.5 --popsFile $popfile --writeFailedWindows --polarize`
